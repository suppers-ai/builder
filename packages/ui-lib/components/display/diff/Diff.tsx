import { BaseComponentProps } from "../../types.ts";

export interface DiffProps extends BaseComponentProps {
  item1Content?: string;
  item2Content?: string;
  item1Label?: string;
  item2Label?: string;
  item1Class?: string;
  item2Class?: string;
}

export function Diff({
  class: className = "",
  item1Content = "BEFORE",
  item2Content = "AFTER", 
  item1Label,
  item2Label,
  item1Class = "bg-primary text-primary-content font-bold text-xl flex items-center justify-center",
  item2Class = "bg-secondary text-secondary-content font-bold text-xl flex items-center justify-center",
  id,
  ...props
}: DiffProps) {
  const diffClasses = [
    "diff",
    "aspect-[16/9]",
    className,
  ].filter(Boolean).join(" ");

  return (
    <div class={diffClasses} id={id} {...props}>
      <div class="diff-item-1">
        <div class={`diff-resizer ${item1Class}`}>
          {item1Label && <div class="text-sm opacity-75 mb-2">{item1Label}</div>}
          {item1Content}
        </div>
      </div>
      <div class="diff-item-2">
        <div class={`diff-resizer ${item2Class}`}>
          {item2Label && <div class="text-sm opacity-75 mb-2">{item2Label}</div>}
          {item2Content}
        </div>
      </div>
    </div>
  );
}

// Keep the individual components for advanced usage
export interface DiffItemProps extends BaseComponentProps {
  children: any;
  item: 1 | 2;
}

export function DiffItem({
  class: className = "",
  children,
  item,
  id,
  ...props
}: DiffItemProps) {
  const itemClasses = [
    `diff-item-${item}`,
    className,
  ].filter(Boolean).join(" ");

  return (
    <div class={itemClasses} id={id} {...props}>
      {children}
    </div>
  );
}

export interface DiffResizerProps extends BaseComponentProps {
  children: any;
}

export function DiffResizer({
  class: className = "",
  children,
  id,
  ...props
}: DiffResizerProps) {
  const resizerClasses = [
    "diff-resizer",
    className,
  ].filter(Boolean).join(" ");

  return (
    <div class={resizerClasses} id={id} {...props}>
      {children}
    </div>
  );
}
