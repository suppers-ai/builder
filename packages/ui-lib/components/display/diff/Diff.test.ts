import { assertEquals, assertStringIncludes } from "https://deno.land/std@0.224.0/assert/mod.ts";
import { assertSnapshot } from "https://deno.land/std@0.224.0/testing/snapshot.ts";
import { renderToString } from "preact-render-to-string";
import { Diff } from "./Diff.tsx";

Deno.test("Diff - basic rendering", () => {
  const html = renderToString(Diff({
    oldContent: "Original text",
    newContent: "Changed text",
  }));
  assertStringIncludes(html, "diff");
  assertStringIncludes(html, "Original text");
  assertStringIncludes(html, "Changed text");
});

Deno.test("Diff - with item1 and item2", () => {
  const html = renderToString(Diff({
    oldContent: "First version",
    newContent: "Second version",
  }));
  assertStringIncludes(html, "diff");
  assertStringIncludes(html, "First version");
  assertStringIncludes(html, "Second version");
});

Deno.test("Diff - responsive behavior", () => {
  const html = renderToString(Diff({
    oldContent: "Mobile view",
    newContent: "Desktop view",
    type: "split",
  }));
  assertStringIncludes(html, "diff");
  assertStringIncludes(html, "diff-item-1");
});

Deno.test("Diff - with custom class", () => {
  const html = renderToString(Diff({
    oldContent: "Original",
    newContent: "Changed",
    class: "custom-diff",
  }));
  assertStringIncludes(html, "custom-diff");
});

Deno.test("Diff - with labels", () => {
  const html = renderToString(Diff({
    oldContent: "Before",
    newContent: "After",
    oldLabel: "Before Changes",
    newLabel: "After Changes",
  }));
  assertStringIncludes(html, "Before Changes");
  assertStringIncludes(html, "After Changes");
});

Deno.test("Diff - split view", () => {
  const html = renderToString(Diff({
    oldContent: "Left side",
    newContent: "Right side",
    split: true,
  }));
  assertStringIncludes(html, "diff");
});

Deno.test("Diff - with resize handle", () => {
  const html = renderToString(Diff({
    oldContent: "Resizable left",
    newContent: "Resizable right",
    showLabels: true,
  }));
  assertStringIncludes(html, "mockup-code");
});

Deno.test("Diff - text comparison", () => {
  const html = renderToString(Diff({
    oldContent: "The quick brown fox",
    newContent: "The quick red fox",
    type: "text",
  }));
  assertStringIncludes(html, "diff");
  assertStringIncludes(html, "brown");
  assertStringIncludes(html, "red");
});

Deno.test("Diff - HTML snapshot", async (t) => {
  const html = renderToString(Diff({
    oldContent: "console.log('Hello World');",
    newContent: "console.log('Hello, DaisyUI!');",
    oldLabel: "Before",
    newLabel: "After",
  }));
  await assertSnapshot(t, html);
});

Deno.test("Diff - HTML snapshot responsive", async (t) => {
  const html = renderToString(Diff({
    oldContent: "Original design mockup",
    newContent: "Updated design mockup",
    type: "split",
    showLabels: true,
  }));
  await assertSnapshot(t, html);
});
