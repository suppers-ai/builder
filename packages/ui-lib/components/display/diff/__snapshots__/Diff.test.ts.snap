export const snapshot = {};

snapshot[`DiffDisplay - HTML snapshot 1`] = `\`<div class="diff text-base mockup-code bg-base-200 relative"><div class="flex"><div class="flex-1 text-center p-2 bg-red-100 text-red-700 font-semibold border-b">Before</div><div class="flex-1 text-center p-2 bg-green-100 text-green-700 font-semibold border-b">After</div></div><div class="space-y-1 p-4"><pre class="bg-red-50 text-red-900 px-2 py-1 rounded border-l-4 border-red-400"><span class="text-red-600 mr-2">-</span>console.log('Hello World');</pre><pre class="bg-green-50 text-green-900 px-2 py-1 rounded border-l-4 border-green-400"><span class="text-green-600 mr-2">+</span>console.log('Hello, daisyUI!');</pre></div></div>\``;

snapshot[`DiffDisplay - HTML snapshot responsive 1`] = `'<div class="diff text-base relative"><div class="diff-item-1"><div class="diff-resizer font-semibold p-2 bg-red-50 text-red-700 border-b border-red-200">Before</div><pre class="diff-resizer bg-red-50 text-red-900 p-4 whitespace-pre-wrap overflow-auto max-h-96">Original design mockup</pre></div><div class="diff-item-2"><div class="diff-resizer font-semibold p-2 bg-green-50 text-green-700 border-b border-green-200">After</div><pre class="diff-resizer bg-green-50 text-green-900 p-4 whitespace-pre-wrap overflow-auto max-h-96">Updated design mockup</pre></div></div>'`;
